libs/arangodb/langchain_arangodb/chains/graph_qa/arangodb.py

```diff
@@ -162,6 +172,187 @@ def from_llm(
             **kwargs,
         )
 
+    def _format_aql(self, aql: str) -> str:
+        keywords = [
+            "WITH",
+            "FOR",
+            "FILTER",
+            "LET",
+            "RETURN",
+            "SORT",
+            "LIMIT",
+            "COLLECT",
+            "INSERT",
+            "UPDATE",
+            "REMOVE",
+            "REPLACE",
+        ]
+        for kw in keywords:
+            # Add newline before keyword if it's not already preceded by one
+            aql = re.sub(rf"\b{kw}\b", f"\n{kw}", aql, flags=re.IGNORECASE)
+        return aql.strip()
+
+    def cache_query(
+        self, queries: Optional[List[Dict[str, str]]] = None
+    ) -> Union[str, List[str]]:
+        """
+        Cache one or more queries generated by the LLM only if
+        they're not already stored.
+
+        If no list is provided, caches the most recent query.
+
+        The query text is always lowercased before storing and checking for duplicates.
+
+        :param queries: Optional list of dicts with 'text' and 'aql' keys.
+        :return: True if a single new entry was inserted,
+                or list of messages for multiple entries.
+        """
+        if self.embedding is None:
+            raise ValueError("Cannot cache queries without an embedding model.")
+
+        collection = self.graph.db.collection(self.query_cache_collection_name)  # type: ignore
+
+        if queries is None:
+            # Fallback: cache the most recent query
+            text = self._last_user_input.strip().lower()  # type: ignore
+            aql = self._last_aql_query
+
+            existing = list(collection.find({"text": text, "aql": aql}, limit=1))
+            if existing:
+                return "This query is already in the cache"
+
+            query_embedding = self.embedding.embed_query(text)  # type: ignore
+            collection.insert(
```

Comment:
for all Query insertions, let's add `_key` to the object we're inserting, using the hash of the Query Text

i.e

```
{
   "_key": self._hash(text),
   "text": text,
   "embedding": ...,
   ...
}
```

Reply:
disregard, added in 3d94f09